File: 148.cpp
#include <iostream>
#include <utility>

using namespace std;

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* sortList(ListNode* head) {
        int len = 0;
        ListNode* tmp = head;
        while (tmp)
        {
            tmp = tmp->next;
            ++len;
        }
        ListNode dummy(-1);
        dummy.next = head;
        for (int i = 1; i < len; i *= 2)
        {
            ListNode* cur = dummy.next;
            ListNode* tail = &dummy;
            while (cur) {
                auto l = cur;
                auto r = split(cur, i);
                cur = split(r, i);
                auto nodes = merge(l, r);
                tail->next = nodes.first;
                tail = nodes.second;
            }
        }
        return dummy.next;
    }

private:
    pair<ListNode*, ListNode*> merge(ListNode* a, ListNode* b) {
        ListNode dummy(-1);
        ListNode *tail = &dummy;
        while (a && b) {
            if (a->val <= b->val) {
                tail->next = a;
                a = a->next;
            }else {
                tail->next = b;
                b = b->next;
            }
            tail = tail->next;
        }
        tail->next = a ? a : b;
        while (tail->next) tail = tail->next;
        return make_pair(dummy.next, tail);
    }

    ListNode* split(ListNode* a, int len) {
        while (len > 1 && a) {
            a = a->next;
            --len;
        }
        ListNode* rest = a ? a->next : nullptr; //有可能从a到结尾不到len个节点，此时a为空
        if (a) a->next = nullptr; //断开左右两部分
        return rest;
    }
};

File: 148.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def sortList(self, head: ListNode) -> ListNode:
        def merge(a, b):
            if not (a and b):
                return a if a else b
            dummy = ListNode(float('-inf'))
            current = dummy
            while a and b:
                if a.val < b.val:
                    current.next = a
                    a = a.next
                else:
                    current.next = b
                    b = b.next
                current = current.next
            current.next = a or b
            return dummy.next

        def split(node):
            if not node or not node.next:
                return node
            a = b = node
            while b.next and b.next.next:
                a = a.next
                b = b.next.next
            mid = a.next
            a.next = None
            return mid
        if not head or not head.next:
            return head
        mid = split(head)
        pre = self.sortList(head)
        post = self.sortList(mid)
        return merge(pre, post)


if __name__ == "__main__":
    s = Solution()
    print()
